compiler_opt/rl/generate_bc_trajectories.py

```diff
@@ -0,0 +1,92 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Module for running compilation and collect data for behavior cloning."""
+
+from typing import Callable, Dict, List, Optional
+
+import numpy as np
+import tensorflow as tf
+from tf_agents.trajectories import policy_step
+from tf_agents.trajectories import time_step
+
+
+class ExplorationWithPolicy:
+  """Policy which selects states for exploration.
+
+  Exploration is fascilitated in the following way. First the policy plays
+  all actions from the replay_prefix. At the following state the policy computes
+  a gap which is difference between the most likely action and the second most
+  likely action according to the randomized exploration policy (distr).
+  If the current gap is smaller than previously maintained gap, the gap is
+  updated and the exploration state is set to the current state.
+  The trajectory is completed by following following the policy from the
+  constructor.
+
+  Attributes:
+    replay_prefix: a replay buffer of actions
+    policy: policy to follow after exhausting the replay buffer
+    explore_policy: randomized policy which is used to compute the gap
+    curr_step: current step of the trajectory
+    explore_step: current candidate for exploration step
+    gap: current difference at explore step between probability of most likely
+      action according to explore_policy and second most likely action
+    explore_on_features: dict of feature names and functions which specify
+      when to explore on the respective feature
+  """
+
+  def __init__(
+      self,
+      replay_prefix: List[np.ndarray],
+      policy: Callable[[time_step.TimeStep], np.ndarray],
+      explore_policy: Callable[[time_step.TimeStep], policy_step.PolicyStep],
+      explore_on_features: Optional[Dict[str, Callable[[tf.Tensor],
+                                                       bool]]] = None,
+  ):
+    self.replay_prefix = replay_prefix
+    self.policy = policy
+    self.explore_policy = explore_policy
+    self.curr_step = 0
+    self.explore_step = 0
+    self.gap = np.inf
+    self.explore_on_features = explore_on_features
+    self._stop_exploration = False
+
+  def advice(self, state: time_step.TimeStep) -> np.ndarray:
+    """Action function for the policy.
+
+    Args:
+      state: current state in the trajectory
+
+    Returns:
+      policy_deca: action to take at the current state.
+
+    """
+    if self.curr_step < len(self.replay_prefix):
+      self.curr_step += 1
+      return np.array(self.replay_prefix[self.curr_step - 1])
+    policy_deca = self.policy(state)
+    distr = tf.nn.softmax(self.explore_policy(state).action.logits).numpy()[0]
+    if not self._stop_exploration and distr.shape[0] > 1 and self.gap > np.abs(
```

Comment:
how do we know distr has 2 values? (distr[0], distr[1])

Reply:
I am checking that there are at least two values and I am taking the first two values. In general we can change this to sort the distribution and compute the gap between the largest and second largest value.

Reply:
leave a comment for posterity that this can be changed? i.e. so we understand "it's a current intentional choice to pick 1st 2, but it's not the only possibility"

Reply:
ExplorationWithPolicy has the following behavior: play actions from the replay_prefix until it is exhausted. Play according to policy after it is exhausted and simultaneously check the conditions for exploration which use explore_policy. If the condition is satisfied at a given step of the trajectory compilation, set explore_step to that step to be used by the exploration logic to take a new decision there. The whole exploration process works in the following way: 1. compile the module with policy and save the trajectory, 2. from 1. we know what step to explore on so construct a new replay prefix which includes all actions before exploration step as played by policy and includes the action selected by the exploration procedure at the exploration step selected in 1. 3. play the replay_prefix from 2. until after the exploration_step, then follow policy until the end. We can make this clear in the current docstring or in the docstring where the exploration logic is contained or maybe more appropriately in the docstring for generate_bc_trajectories.py in the beginning.

Reply:
Done.